File: typescript/packages/core/src/commands/builtin/generic/wc.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import { resolveSource } from '../utils/stream.ts'
import { formatRecords } from '../utils/output.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

type Stream = (p: PathSpec) => AsyncIterable<Uint8Array>

async function* wcLinesStream(source: AsyncIterable<Uint8Array>): AsyncIterable<Uint8Array> {
  let count = 0
  for await (const chunk of source) {
    for (let i = 0; i < chunk.byteLength; i++) if (chunk[i] === 0x0a) count += 1
  }
  yield ENC.encode(`${String(count)}\n`)
}

function countChar(text: string, ch: string): number {
  let n = 0
  for (const c of text) if (c === ch) n += 1
  return n
}

export interface WcRow {
  values: number[]
  label: string | null
}

// GNU wc layout: counts right-aligned to a shared width and space-separated;
// a single count for a single operand prints unpadded, and a default-mode
// stdin read uses GNU's width 7 for unknown sizes. Divergence from GNU: the
// width is the widest printed number, while GNU derives it from operand file
// sizes; the two are identical in the default mode, where the byte count is
// the widest column.
export function formatWcLines(rows: WcRow[]): string[] {
  const first = rows[0]
  if (rows.length === 1 && first?.values.length === 1) {
    const body = String(first.values[0])
    return [first.label === null ? body : `${body} ${first.label}`]
  }
  let width = 1
  if (rows.length === 1 && first?.label === null) {
    width = 7
  } else {
    for (const row of rows) {
      for (const n of row.values) width = Math.max(width, String(n).length)
    }
  }
  return rows.map((row) => {
    const body = row.values.map((n) => String(n).padStart(width)).join(' ')
    return row.label === null ? body : `${body} ${row.label}`
  })
}

export async function wcGeneric(
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
  stream: Stream,
): Promise<CommandFnResult> {
  const f = opts.flags
  const lFlag = f.args_l === true
  const wFlag = f.w === true
  const cFlag = f.c === true
  const mFlag = f.m === true
  const LFlag = f.L === true
  if (paths.length > 0) {
    const rows: WcRow[] = []
    let totalLines = 0
    let totalWords = 0
    let totalBytes = 0
    let totalMax = 0
    for (const p of paths) {
      const data = await materialize(stream(p))
      const text = DEC.decode(data)
      const lineCount = countChar(text, '\n')
      const wordCount = text.split(/\s+/).filter((s) => s !== '').length
      const byteCount = data.byteLength
      if (LFlag) {
        const maxLen = text.split(/\r?\n/).reduce((m, l) => Math.max(m, l.length), 0)
        rows.push({ values: [maxLen], label: p.original })
        totalMax = Math.max(totalMax, maxLen)
      } else if (lFlag) {
        rows.push({ values: [lineCount], label: p.original })
        totalLines += lineCount
      } else if (wFlag) {
        rows.push({ values: [wordCount], label: p.original })
        totalWords += wordCount
      } else if (cFlag) {
        rows.push({ values: [byteCount], label: p.original })
        totalBytes += byteCount
      } else if (mFlag) {
        const charCount = text.length
        rows.push({ values: [charCount], label: p.original })
        totalBytes += charCount
      } else {
        rows.push({ values: [lineCount, wordCount, byteCount], label: p.original })
        totalLines += lineCount
        totalWords += wordCount
        totalBytes += byteCount
      }
    }
    if (paths.length > 1) {
      if (LFlag) rows.push({ values: [totalMax], label: 'total' })
      else if (lFlag) rows.push({ values: [totalLines], label: 'total' })
      else if (wFlag) rows.push({ values: [totalWords], label: 'total' })
      else if (cFlag || mFlag) rows.push({ values: [totalBytes], label: 'total' })
      else rows.push({ values: [totalLines, totalWords, totalBytes], label: 'total' })
    }
    const out: ByteSource = formatRecords(formatWcLines(rows))
    return [out, new IOResult()]
  }
  let source: AsyncIterable<Uint8Array>
  try {
    source = resolveSource(opts.stdin, 'wc: missing operand')
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
  }
  if (lFlag) return [wcLinesStream(source), new IOResult()]
  const raw = await materialize(source)
  const text = DEC.decode(raw)
  const lc = countChar(text, '\n')
  const wcVal = text.split(/\s+/).filter((s) => s !== '').length
  const bc = raw.byteLength
  const cc = text.length
  if (LFlag) {
    const maxLen = text.split(/\r?\n/).reduce((m, l) => Math.max(m, l.length), 0)
    return [ENC.encode(`${String(maxLen)}\n`), new IOResult()]
  }
  if (wFlag) return [ENC.encode(`${String(wcVal)}\n`), new IOResult()]
  if (mFlag) return [ENC.encode(`${String(cc)}\n`), new IOResult()]
  if (cFlag) return [ENC.encode(`${String(bc)}\n`), new IOResult()]
  const line = formatWcLines([{ values: [lc, wcVal, bc], label: null }])[0] ?? ''
  return [ENC.encode(`${line}\n`), new IOResult()]
}

File: python/mirage/commands/builtin/paperclip/wc.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.accessor.paperclip import PaperclipAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.utils.stream import _read_stdin_async
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.paperclip.glob import resolve_glob
from mirage.core.paperclip.read import read as paperclip_read
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("wc", resource="paperclip", spec=SPECS["wc"])
async def wc(
    accessor: PaperclipAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    args_l: bool = False,
    w: bool = False,
    c: bool = False,
    m: bool = False,
    L: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if paths:
        paths = await resolve_glob(accessor, paths)
        p = paths[0]
        data = await paperclip_read(accessor, p, index)
    else:
        data = await _read_stdin_async(stdin)
        if data is None:
            raise ValueError("wc: missing operand")
    text = data.decode(errors="replace")
    line_count = text.count("\n")
    word_count = len(text.split())
    byte_count = len(data)
    if L:
        max_len = max((len(ln) for ln in text.splitlines()), default=0)
        return str(max_len).encode(), IOResult()
    if args_l:
        return str(line_count).encode(), IOResult()
    if w:
        return str(word_count).encode(), IOResult()
    if m:
        return str(len(text)).encode(), IOResult()
    if c:
        return str(byte_count).encode(), IOResult()
    return f"{line_count}\t{word_count}\t{byte_count}".encode(), IOResult()
